homeassistant/components/proximity.py

```diff
@@ -0,0 +1,325 @@
+
+    _LOGGER.debug('zone settings: LAT:%s LONG:%s', proximity_latitude,
+                  proximity_longitude)
+
+    # create an entity so that the proximity values can be used for other
+    # components
+    entities = set()
+
+    # set the default values
+    dist_to_zone = 'not set'
+    dir_of_travel = 'not set'
+    nearest = 'not set'
+
+    proximity = Proximity(hass, zone_friendly_name, dist_to_zone,
+                          dir_of_travel, nearest)
+    proximity.entity_id = entity_id
+
+    proximity.update_ha_state()
+    entities.add(proximity.entity_id)
+
+    def check_proximity_state_change(entity, old_state, new_state):
+        # pylint: disable=too-many-branches,too-many-statements,too-many-locals
+        """ Function to perform the proximity checking """
+        entity_name = new_state.attributes['friendly_name']
+        device_is_in_zone = False
+        devices_to_calculate = False
+        devices_in_zone = ''
+
+        # check for devices in the monitored zone
+        for device in proximity_devices:
+            device_state = hass.states.get(device)
+
+            if device_state.state not in ignored_zones:
+                devices_to_calculate = True
+
+            # check the location of all devices
+            if device_state.state == config[DOMAIN]['zone']:
+                device_is_in_zone = True
+                device_friendly = device_state.attributes['friendly_name']
+                if devices_in_zone != '':
+                    devices_in_zone = devices_in_zone + ', '
+                devices_in_zone = devices_in_zone + device_friendly
+                _LOGGER.info('%s: %s is in the monitored zone: %s',
+                             entity_name, device, device_state.state)
+
+        # no-one to track so reset the entity
+        if not devices_to_calculate:
+            proximity.dist_to = 'not set'
+            proximity.dir_of_travel = 'not set'
+            proximity.nearest = 'not set'
+            proximity.update_ha_state()
+            _LOGGER.debug('%s: all devices in an ignored zone', entity_name)
+            return
+
+        # at least one device is in the monitored zone so update the entity
+        if device_is_in_zone:
+            proximity.dist_to = 0
+            proximity.dir_of_travel = 'arrived'
+            proximity.nearest = devices_in_zone
+            proximity.update_ha_state()
+            _LOGGER.debug('%s: update entity: distance=0: direction='
+                          'arrived: device=%s', entity_name, devices_in_zone)
+            return
+
+        # we can't check proximity because latitude and longitude don't exist
+        if 'latitude' not in new_state.attributes:
+            _LOGGER.info('%s: not LAT or LONG current position cannot be '
+                         'calculated', entity_name)
+            return
+
+        # collect distances to the zone for all devices
+        distances_to_zone = {}
```

Comment:
Soo, you could fill this dict with distances with a dict comprehension (untested):

```python
distances_to_zone = {
  state.entity_id: round(distance(…)/1000, 1) for state in
    (hass.states.get(entity_id) for entity_id in proximity_devices)
  if state is not None and state.state not in ignored_zones and ATTR_LATITUDE in state.attributes
}

closest_device = min(distances_to_zone.items(), min=lambda item: item[1])[0]
```